### PR TITLE
Migrate the aging goamz library to the latest aws-sdk

### PR DESCRIPTION
**Pull request by lox**
This is take 2 of https://github.com/buildkite/agent/pull/185. 

It aims to be a minimal update to the latest aws-sdk. There are lots of nice things that can be done with parallel uploads and downloads, but in the meantime there are bug fixes that are needed.

* [x] Update ec2-metadata
* [x] Update ec2-tags
* [x] Update s3-uploader
* [x] Update s3-downloader

### CODE REVIEW

**Comment by keithpitt** [comment]
@lox is this still a WIP? Or ready for a review party?

**Comment by lox** [comment]
Pretty sure it's actually ready.

File: agent/s3.go [hunk 1]
```diff
@@ -2,75 +2,109 @@ package agent
...
+	e.retrieved = true
 	return
 }
 
-func awsS3Region() (region aws.Region, err error) {
+func (e *credentialsProvider) IsExpired() bool {
+	return !e.retrieved
+}
+
+func awsS3Credentials() *credentials.Credentials {
```
Comment (keithpitt) [inline agent/s3.go]: Does this support the same ENV vars as this?

https://github.com/goamz/goamz/blob/master/aws/aws.go#L340

Comment (lox) [inline agent/s3.go]: Yup! https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/#EnvProvider

File: agent/s3.go [hunk 2]
```diff
@@ -2,75 +2,109 @@ package agent
...
+	return credentials.NewChainCredentials(
+		[]credentials.Provider{
+			&credentialsProvider{},
+			&credentials.EnvProvider{},
+			&ec2rolecreds.EC2RoleProvider{},
+		})
+}
+
+func awsS3RegionFromEnv() (region string, err error) {
 	regionName := "us-east-1"
```
Comment (lox) [inline agent/s3.go]: I reckon this should probably be based on the region of the bucket in question.

File: agent/s3.go (hunk 1 above)
Comment (keithpitt) [inline agent/s3.go]: Radical.

File: agent/s3.go (hunk 2 above)
Comment (keithpitt) [inline agent/s3.go]: How can you tell a region based on a bucket name?

Comment (lox) [inline agent/s3.go]: https://docs.aws.amazon.com/sdk-for-go/api/service/s3/s3manager/#GetBucketRegion I assume magic?

Comment (keithpitt) [inline agent/s3.go]: > GetBucketRegion will attempt to get the region for a bucket using the regionHint to determine which AWS partition to perform the query on.
> 
> The request will not be signed, and will not use your AWS credentials.

That's some cool sounding magic! Yeah, I reckon give that a go and if it doesn't work, fall back to the default.

**Comment by lox** [comment]
I've given this an initial test against Minio and it works well. Next up is some testing against S3.

**Comment by muhlba91** [comment]
@lox i am struggling to find out how to configure the artifacts storage for a private minio bucket where i need to set the endpoint url. would you be able to provide me an example on how i'd need to set the configuration for this use case as you tested against minio?

**Pull request merged**